### ISSUE TITLE
FIDO U2F (Yubico Yubikeys)

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#3043]
I've typically pointed people to [QTBUG-54720](https://bugreports.qt.io/browse/QTBUG-54720) for FIDO U2F support, but it might be possible for qutebrowser to implement this independently from QtWebEngine. Some pointers:

- [python-u2flib-host](https://github.com/Yubico/python-u2flib-host)
- [u2f-chrome-extension](https://github.com/google/u2f-ref-code/tree/master/u2f-chrome-extension) (reference implementation)
- [u2f4moz](https://github.com/prefiks/u2f4moz)
- [pyu2f](https://github.com/google/pyu2f)

**Comment by dummys** [comment]
+1 on this issue, it's a killer feature!

**Comment by McKean** [comment]
@The-Compiler are you aware of any way to use this without browser integration?

**Comment by The-Compiler** [comment]
Nope

**Comment by gabm** [comment]
that would be a great feature for me as well...

**Comment by teon** [comment]
This is the only feature that's missing for me..

**Comment by The-Compiler** [comment]
You might want to use the [reactions feature](https://github.com/blog/2119-add-reactions-to-pull-requests-issues-and-comments) instead of adding comments :wink:.

I still doubt whether qutebrowser can support this without it being implemented (in C++) in QtWebEngine. IIRC there have been some attempts at making this work, but nothing came out of it as qutebrowser can't run any privileged JavaScript. There might be ways to get something to work by bridging JavaScript and C++/Python, but I don't see that happening anytime soon unless someone steps up with a working proof-of-concept.

**Comment by McKean** [comment]
Perhaps this helps for some inspiration: https://github.com/brave/browser-laptop/issues/518

**Comment by dijit** [comment]
The brave example is an electron application. Which is fundamentally different from python+QtWebEngine.

Sadly.

**Comment by The-Compiler** [comment]
According to https://github.com/electron/electron/issues/3226#issuecomment-433291750 U2F support is built into Chromium starting with Chrome 69, which means Qt 5.12. There's some small chance it'll just work™ there, or if not, it should get much easier for QtWebEngine to implement it in the future.

Does anyone here have a Qt 5.12 to test, and a U2F token available? I bought one (when they were cheap thanks to some GitHub promotion) just to test it with qutebrowser, but I have no idea where that one went, as I never used it since :laughing:

**Comment by mschilli87** [comment]
@The-Compiler:
> Does anyone here have a Qt 5.12 to test, and a U2F token available?

My girlfriend has one, so I *should* be able to test once you have something you want tested. I'd need to  build Qt 5.12 (should be possible) but have no clue how to setup/test the U2F token.

**Comment by oliversturm** [comment]
@The-Compiler I have a U2F token that is set up to work with various sites in Chrome. However, I don't have Qt 5.12 built yet. I'll try to do that asap and let you know if it changes anything in current qutebrowser.

**Comment by The-Compiler** [comment]
FWIW a Qt developer answered that it's unlikely this works as they haven't WebUSB support enabled so far. Still, if someone wants to check, you can probably do so here: https://demo.yubico.com/u2f

**Comment by Ulrar** [comment]
You shouldn't need webusb for U2F

**Comment by McKean** [comment]
FWIW it doesn't seem to be working.
![screenshot](https://dl.dropbox.com/s/dbq0pfjwsraflgc/shot-2018-10-29-11-58-54.png?dl=0)

**Comment by jeremytregunna** [comment]
I do have 5.12, and can confirm it does not work on any U2F token (be it YubiKey or my own self-built one).

**Comment by marza91** [comment]
Hey, I just got me a Trezor model T, and for me the connection seem to work in Trezors own web frontend, and U2F in Dropbox. Google does not (doesn't discover the device), and Github tells me the browser doesn't support the FIDO U2F standard. I have the official Trezor Bridge daemon installed which is what Trezor themselves use, so I suspect Dropbox does the same thing?
Running the latest Arch Linux.

**Comment by machin3io** [comment]
It's a real bummer this isn't working.

**Comment by McKean** [comment]
I believe we just have to sit tight as already mentioned. Now that there is a defined standard I'm guessing it will not take too long until qtwebengine supports webauthn (https://www.w3.org/TR/webauthn/).

**Comment by ashley-woodard** [comment]
Signs of [progress](https://demo.yubico.com/u2f) running `qt5-webengine-5.12.4`:

I fully expected this to fail...

![image](https://user-images.githubusercontent.com/70843/60154049-c877b180-97ab-11e9-80b7-b2fc7493635e.png)

But surely this won't work...

![image](https://user-images.githubusercontent.com/70843/60154057-d299b000-97ab-11e9-9b95-5e6d6fa959ad.png)

Seems legit... >_>

![image](https://user-images.githubusercontent.com/70843/60154071-dd544500-97ab-11e9-86ce-93771937a0e7.png)

YMMV and all that jazz.

**Comment by ashley-woodard** [comment]
I am pleased to report that twitter, qutebrowser, and the YubiKey 5 NFC works like a charm as far as U2F is concerned. AFAICT, twitter is simply one of the few major sites refraining from explicit checks for Chrome. Perhaps if other sites would stop essentially _blacklisting_ other browsers, everyone could profit -- not just esoteric vim-inspired browsers, etc, et al. Bah.

**Comment by gabm** [comment]
thats good news, I can confirm the yubico test-site works.. gitlab doesn't though, even faking the user-agent string doesnt help...

**Comment by IvanDelsinne** [comment]
GitHub doesn't seem to accept U2F authentication either, which is pretty annoying.

**Comment by PSC54** [comment]
@ashley-woodard 
Can you give us a small tutorial how you get U2F working, please? 
I always get this Error: 
![2019-09-02-111705_1366x768_scrot](https://user-images.githubusercontent.com/44315232/64107166-2ad7ca00-cd7a-11e9-9454-542ea6b57e0f.png)

I using arch, qt5-webengine-5.13.0-4. I tried the official qutebrowser from pacman and the qutebrowser-git from the AUR. 
I also using different user agents without success.

**Comment by jb55** [comment]
I'm on 5.12.1 and it works, but my colleagues running 5.13 says it doesn't, perhaps something broke?

**Comment by PSC54** [comment]
> I'm on 5.12.1 and it works, but my colleagues running 5.13 says it doesn't, perhaps something broke?

What qutebrowser version do you use?

Downgrading qt5-webengine is not so easy. Get some trouble with other libs when downgrade to 5.12.1 and qutebrowser does not start.

**Comment by The-Compiler** [comment]
@PSC54 You could use a [tox install](https://github.com/qutebrowser/qutebrowser/blob/master/doc/install.asciidoc#tox) with an older version.

**Comment by jb55** [comment]
@PSC54 I'm running 1.7.0:

```
qutebrowser v1.7.0
Git commit: 
Backend: QtWebEngine (Chromium 69.0.3497.128)

CPython: 3.7.4
Qt: 5.12.3
PyQt: 5.13.0

sip: 4.19.18
colorama: no
pypeg2: 2.15
jinja2: 2.10.1
pygments: 2.4.2
yaml: 5.1.1
cssutils: 1.0.2 $Id$
attr: 18.2.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebKitWidgets: no
pdf.js: 2.1.266 (/nix/store/kyiyfrrky1b73js8b4g4wnws951pi66g-pdfjs-2.1.266/build/pdf.js)
sqlite: 3.28.0
QtNetwork SSL: OpenSSL 1.1.1c  28 May 2019

Style: QFusionStyle
Platform: Linux-5.3.0-rc8-x86_64-with-glibc2.2.5, 64bit
Linux distribution: NixOS 19.09.git.b6dcdb1 (Loris) (unknown)
Frozen: False
Imported from /nix/store/yvm6kvqw8nb421l8r3zpi125mppp2j0z-qutebrowser-1.7.0/lib/python3.7/site-packages/qutebrowser
Using Python from /nix/store/h6ca7b9k0zmw71ay1q7iq28g08r1i8ji-python3-3.7.4/bin/python3.7
Qt library executable path: /nix/store/6svq4r1qiahp4fsgh2893zl6pzcl1xvc-qtbase-5.12.3/libexec, data path: /nix/store/6svq4r1qiahp4fsgh2893zl6pzcl1xvc-qtbase-5.12.3
OS Version: 

--- /etc/os-release ---
NAME=NixOS
ID=nixos
VERSION="19.09.git.b6dcdb1 (Loris)"
VERSION_CODENAME=loris
VERSION_ID="19.09.git.b6dcdb1"
PRETTY_NAME="NixOS 19.09.git.b6dcdb1 (Loris)"
LOGO="nix-snowflake"
DOCUMENTATION_URL="https://nixos.org/nixos/manual/index.html"
```

**Comment by jb55** [comment]
> I'm on 5.12.1 and it works, but my colleagues running 5.13 says it doesn't, perhaps something broke?

oops I meant 5.12.3

**Comment by apoelstra** [comment]
I am @jb55's colleague. Can confirm that the Yubikey does not work with the Arch repo version of qutebrowser (which uses Qt 5.13) but **does** work with the tox-installed version (which use 5.12, apparently due to an unrelated bug).

**Comment by instagibbs** [comment]
Also @jb55 colleague, can confirm tox steps work for ubuntu 18.04 LTS

**Comment by McKean** [comment]
I'm one of the unlucky ones, doesn't work :(
```
qutebrowser v1.8.1
Git commit: 
Backend: QtWebEngine (Chromium 73.0.3683.105)

CPython: 3.7.4
Qt: 5.13.1
PyQt: 5.13.1
```

**Comment by dfischer** [comment]
```
qutebrowser v1.8.1
Git commit: a733fa9fd (2019-11-18 16:40:15 +0100)
Backend: QtWebEngine (Chromium 69.0.3497.128)

CPython: 3.7.5rc1
Qt: 5.12.5 (compiled 5.12.4)
PyQt: 5.12.3

sip: 5.0.0
colorama: 0.4.1
pypeg2: 2.15
jinja2: 2.10.3
pygments: 2.4.2
yaml: 5.1.2
cssutils: 1.0.2 $Id$
attr: 19.3.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: yes
PyQt5.QtWebKitWidgets: no
pdf.js: no
sqlite: 3.28.0
QtNetwork SSL: OpenSSL 1.1.1c  28 May 2019

Style: QFusionStyle
Platform: Linux-5.3.0-22-generic-x86_64-with-Ubuntu-19.10-eoan, 64bit
Linux distribution: Pop!_OS 19.10 (ubuntu)
Frozen: False
Imported from /home/dfischer/.build/qutebrowser/qutebrowser
Using Python from /home/dfischer/.build/qutebrowser/.venv/bin/python3
Qt library executable path: /home/dfischer/.build/qutebrowser/.venv/lib/python3.7/site-packages/PyQt5/Qt/libexec, data path: /home/dfischer/.build/qutebrowser/.venv/lib/python3.7/site-packages/PyQt5/Qt

Paths:
cache: /home/dfischer/.cache/qutebrowser
config: /home/dfischer/.config/qutebrowser
data: /home/dfischer/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser

Autoconfig loaded: yes
Config.py: no config.py was loaded
Uptime: 0:00:00

```

**Comment by PhoenixBurns** [comment]
Is this something forthcoming?  Would love to see it in qute since I had to use firefox to log in and post this.

**Comment by The-Compiler** [comment]
@PhoenixBurns did you read the recent comments?

**Comment by klardotsh** [comment]
On Gentoo with Qt 5.13, no dice on the Yubico demo page nor on Github.

```
qutebrowser v1.8.3
Git commit:
Backend: QtWebEngine (Chromium 73.0.3683.105)

CPython: 3.6.9
Qt: 5.13.2
PyQt: 5.12.2

sip: 4.19.17
colorama: 0.3.3
pypeg2: 2.15
jinja2: 2.10.1
pygments: 2.4.2
yaml: 5.1
cssutils: no
attr: 19.3.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: yes
PyQt5.QtWebKitWidgets: yes
pdf.js: no
sqlite: 3.30.1
QtNetwork SSL: LibreSSL 3.0.2

Style: QWindowsStyle
Platform: Linux-5.4.6-gentoo-x86_64-x86_64-AMD_Ryzen_7_2700X_Eight-Core_Processor-with-gentoo-2.6, 64bit
Linux distribution: Gentoo/Linux (gentoo)
Frozen: False
Imported from /usr/lib64/python3.6/site-packages/qutebrowser
Using Python from /usr/bin/python3.6
Qt library executable path: /usr/lib64/qt5/libexec, data path: /usr/share/qt5

Paths:
cache: /home/j/.cache/qutebrowser
config: /home/j/.config/qutebrowser
data: /home/j/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser

Autoconfig loaded: yes
Config.py: no config.py was loaded
Uptime: 0:00:00
```

**Comment by gordon-quad** [comment]
Any clues why it works on some systems but doesn't work on others?
Some qtwebengine build flags differ perhaps?
Have no luck with 5.14.2 on Gentoo. Any reports for it working on 5.14.2 anywhere else?

**Comment by The-Compiler** [comment]
I'm afraid I have no idea...

**Comment by terlar** [comment]
I have it working on QtWebEngine 5.15.0

And for me it has been working for the past months as well, so suspect that version worked as well. I remember it was working and then stopped working and then started working again sometime during last year (hence me following this issue). Not sure what made it work, it can also be different sites having an allow list on which browsers they enable FIDO U2F to. GitHub was my example that stopped working for a while.

**Comment by McKean** [comment]
> it can also be different sites having an allow list on which browsers they enable FIDO U2F

I highly doubt this.
On github for example I can try to use my key but it fails. Does it work for you on github now @terlar? 

```
qutebrowser v1.12.0
Git commit: 
Backend: QtWebEngine (Chromium 80.0.3987.163)
Qt: 5.15.0

CPython: 3.8.3
PyQt: 5.15.0
```

**Comment by terlar** [comment]
Yup, it is working now.

```
qutebrowser v1.13.0
Git commit: 
Backend: QtWebEngine (Chromium 69.0.3497.128)
Qt: 5.12.7

CPython: 3.8.3
PyQt: 5.14.2
```

**Comment by winny-** [comment]
No dice with (from portage with a bumped qutebrowser package):

```
qutebrowser v1.13.0
Git commit: 
Backend: QtWebEngine (Chromium 80.0.3987.163)
Qt: 5.15.0

CPython: 3.7.8
PyQt: 5.15.0
```

Just to be clear, this page works for you @tartar: https://demo.yubico.com/webauthn-technical/registration

**Comment by gordon-quad** [comment]
@terlar can I ask which OS/distribution are you running and qtwebengine package version as well? I want to take a peek on how it is built there

**Comment by winny-** [comment]
As @terlar suggested it did work via nixpkgs. I just installed qutebrowser from nixpkgs-unstable and was able to use my u2f device (via [this page](https://demo.yubico.com/webauthn-technical/registration)).

([full version info](https://gist.github.com/winny-/feb5d83683340a1793459c7561b342b2))

```
qutebrowser v1.13.0
Git commit: 
Backend: QtWebEngine (Chromium 69.0.3497.128)
Qt: 5.12.7

CPython: 3.8.3
PyQt: 5.14.2
```

Steps to install on non-nix is a bit wonky. Steps for gentoo, most distros should be the same after step 2:

1. Install nix `layman -a nix-guix && emerge -av sys-apps/nix`
2. Enable the nix daemon `rc-update add nix-daemon default && rc-service nix-daemon start`
3. Relogin or `. /etc/profile`
4. Add the nixpkgs-unstable channel `nix-channel --add https://nixos.org/channels/nixpkgs-unstable`
5. Refresh the channels `nix-channel --update`
6. Install qutebrowser `nix-env -i qutebrowser`
7. Install [nixGL](https://github.com/guibou/nixGL) kludge to uses system libgl with qutebrowser
  a. `git clone https://github.com/guibou/nixGL && cd nixGL`
  b. `nix-env -f ./ -iA nixGLDefault`
8. Run `nixGL qutebrowser` (using `-T` if you already have a qutebrowser window open)

**Comment by terlar** [comment]
Yes, I am running NixOS and as I suspected it would work for anyone installing it via Nix, I would be very surprised otherwise.

You can easily find how it is build on Nix:
https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/networking/browsers/qutebrowser/default.nix

As long as you have Nix setup the steps could be simplified a bit:
w/o. NixGL
1. `nix-env -f https://github.com/NixOS/nixpkgs/archive/nixos-unstable.tar.gz -iA qutebrowser`
2. `qutebrowser`

w. NixGL
1. `nix-env -f https://github.com/NixOS/nixpkgs/archive/nixos-unstable.tar.gz -iA qutebrowser`
2. `nix-env -f https://github.com/guibou/nixGL/archive/master.tar.gz -iA nixGLDefault`
3. `nixGL qutebrowser`

**Comment by The-Compiler** [comment]
FWIW there's some traction upstream now, after someone with Qt Premium Support apparently requested this upstream. See [Support building with enable_web_auth=true (Ia68303fc) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine-chromium/+/308181).

**Comment by The-Compiler** [comment]
The upstream issue has now been closed, with U2F support apparently being enabled for Qt 5.15.1. I haven't tested it myself yet, but I suppose this can finally be closed with no changes from qutebrowser's side.

Qt 5.15.1 is planned for August, so I'll still wait until it's out and someone confirms that it works.

**Comment by mschilli87** [comment]
Once it is available for testing in Gentoo and I find the time to build it, I can try it out and report back. So feel free to ping me if this doesn't happen and none of the Arch users test it either. :wink:

**Comment by twouters** [comment]
I've applied the patches on 5.15.0 (Arch Linux) and can confirm that U2F works on https://demo.yubico.com/webauthn-technical/registration, demo playground and have successfully registered my key on my github account.

Have not been able to get it to work on Gitlab.

Not sure if relevant but user agent is set to: `content.headers.user_agent = Mozilla/5.0 ({os_info}) AppleWebKit/{webkit_version} (KHTML, like Gecko) {qt_key}/{qt_version} {upstream_browser_key}/{upstream_browser_version} Safari/{webkit_version}`

**Comment by jpalus** [comment]
Also can confirm that FastMail, GMail and GitHub are working fine now with 2FA.

**Comment by The-Compiler** [comment]
Closing this then! I don't think it's worthwhile (and I doubt it's possible at all!) to implement U2F directly in qutebrowser.

**Comment by The-Compiler** [comment]
A couple of related feature requests upstream:

- [[QTBUG-90938] Support FIDO2 user verification - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-90938)
- [[QTBUG-90941] Support FIDO2 resident credentials - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-90941)

Thanks to @maximbaz for reporting those!

**Comment by sudoforge** [comment]
Commenting here for anybody else encountering this issue:

Yubikeys don't appear to work at all on Okta's second-factor page while using Qutebrowser.

**Comment by patrick-kidger** [comment]
What's the current status of this issue? I'm trying a Yubikey with versions
```
qutebrowser v3.1.0
Backend: QtWebEngine 6.6.1, based on Chromium 112.0.5615.213 (from api)
Qt: 6.6.1
```
and not getting any kind of prompt when trying to set this up on a Google account.

(Do I understand from above that this might still not work out-of-the-box?)

**Comment by twouters** [comment]
> What's the current status of this issue? I'm trying a Yubikey with versions
> 
> ```
> qutebrowser v3.1.0
> Backend: QtWebEngine 6.6.1, based on Chromium 112.0.5615.213 (from api)
> Qt: 6.6.1
> ```
> 
> and not getting any kind of prompt when trying to set this up on a Google account.
> 
> (Do I understand from above that this might still not work out-of-the-box?)

Works for me out of the box.
To confirm: visit https://demo.yubico.com/webauthn-technical/registration and follow the instructions.

**Comment by jpalus** [comment]
Works fine for me for years including Google.

Note that `qutebrowser` does not show any prompts by itself. Sometimes web pages give a clue that action on security key is needed but not always. Google if I remember correctly does not. But I can tell looking at my YubiKey 5 with flashing light that I need to touch it.

**Comment by sudoforge** [comment]
> What's the current status of this issue? I'm trying a Yubikey with versions
> ```
> qutebrowser v3.1.0
> Backend: QtWebEngine 6.6.1, based on Chromium 112.0.5615.213 (from api)
> Qt: 6.6.1
> ```
> and not getting any kind of prompt when trying to set this up on a Google account.
> 
> (Do I understand from above that this might still not work out-of-the-box?)

Security keys have worked out of the box for a while now. My last comment in 2021 was caused by a failure on a specific application, but IIRC this was due to the application developer and resolved itself some time later.

As has been mentioned, you won't see a modal or desktop prompt; qutebrowser emits no notice that an application is waiting for you to tap your security key (although some applications do update their UI to indicate this; e.g. GitHub).

If you're encountering a problem or think you've found a potential bug using your security key(s), please open a new issue.

**Comment by patrick-kidger** [comment]
Aha, indeed it turned out this was working all along, I just didn't see any prompt and assumed it wasn't working.
Thank you everyone for the help, I really appreciate it!
